src/serialization/deserializer.rs

```diff
@@ -266,10 +265,12 @@ impl <T: Counter> DecodeLoopState<T> {
             let count: T = T::from_i64(count_or_zeros)
                 .ok_or(DeserializeError::UnsuitableCounterType)?;
 
-            h.set_count_at_index(self.dest_index, count)
-                .map_err(|_| DeserializeError::EncodedArrayTooLong)?;
+            if count > T::zero() {
+                h.set_count_at_index(self.dest_index, count)
+                    .map_err(|_| DeserializeError::EncodedArrayTooLong)?;
 
-            restat_state.on_nonzero_count(self.dest_index, count);
+                restat_state.on_nonzero_count(self.dest_index, count);
```

Comment:
You'd think the `on_nonzero_count` would have been enough of a clue for me when writing the original logic, but apparently not. :/